range_queries/heavy_light_decomposition.cpp

```diff
@@ -0,0 +1,578 @@
+		for(int k = 1; k < t_maxlift; k++) {
+			if(t_par[u][k-1] != -1) {
+				t_par[u][k] = t_par[t_par[u][k-1]][k-1];
+			}
+		}
+
+		for(const int& v: t_adj[u]) {
+			if(v^p) {
+				dfs_lca(v, u);
+			}
+		}
+	}
+
+public:
+	/**
+	 * @brief Class parameterized constructor, resizes the and initializes the data members
+	 * @param nodes the total number of nodes in the tree
+	*/
+	explicit Tree(int nodes) 
+		: t_nodes(nodes), t_maxlift(static_cast<int>(floor(log2(nodes))) + 1) {
+		/* Initialize and resize all the vectors */
+		t_root = 0; /* Default */
+		t_adj.resize(t_nodes);
+		t_par.assign(t_nodes, std::vector<int>(t_maxlift, -1));
+		t_depth.assign(t_nodes, 0);
+		t_size.assign(t_nodes, 1);
+		t_val.resize(t_nodes);
+	}
+	
+	/**
+	 * @brief Adds an undirected edge from node u to node v in the tree
+	 * @param u the node where the edge is from
+	 * @param v the node where the edge is to
+	*/
+	void add_edge(const int u, const int v) {
+		t_adj[u].push_back(v);
+		t_adj[v].push_back(u);
+	}
+
+	/**
+	 * @brief Set the root for the tree
+	 * @param new_root the new root
+	*/
+	void change_root(int new_root) {
+		t_root = new_root;
+	}
+	
+	/**
+	 * @brief Set the values for all the nodes
+	 * @param node_val a vector of size n, with all the node values where, n is the number of nodes
+	*/
+	void set_node_val(std::vector<X> node_val) {
+		assert(static_cast<int>(node_val.size()) == t_nodes);
+		t_val = node_val;
+	}
+
+	/**
+	 * @brief This function must be called after the tree adjacency list and node values are populated
+	 * The function initializes the required parameters, and populates the segment tree
+	*/
+	void init() {
+		assert(t_nodes > 0);
+		dfs_size(t_root);
+		dfs_lca(t_root);
+	}
+
+	/**
+	 * @brief The function lifts a node, k units up the tree.
+	 * The lifting is done in place, and the result is stored in the address pointed by p.
+	 * @param p a pointer to the variable that stores the node id
+	 * @param dist the distance to move up the tree
+	 * @returns void.
+	*/
+	void lift(int* const p, int dist) {
+		for(int k = 0; k < t_maxlift; k++) {
+			if(*p == -1) return;
+			if(dist & 1) {
+				*p = t_par[*p][k];
+			}
+			dist >>= 1;
+		}
+	} 
+	
+	/**
+	 * @brief The function returns the kth ancestor of a node
+	 * @param p the node id whose kth ancestor is to be found
+	 * @param dist the distance to move up the tree
+	 * @returns the kth ancestor of node
+	*/
+	int kth_ancestor(int p, const int& dist) {
+		lift(&p, dist);
+		return p;
+	}
+
+	/**
+	 * @brief The function returns the least common ancestor of two nodes
+	 * @param a node id_1
+	 * @param b node id_2
+	 * @returns the least common ancestor of node a, and node b
+	*/
+	int lca(int a, int b) {
+		assert(a >= 0 and b >= 0 and a < t_nodes and b < t_nodes);
+		if(t_depth[a] > t_depth[b]) {
+			lift(&a, t_depth[a] - t_depth[b]);
+		}
+		if(t_depth[b] > t_depth[a]) {
+			lift(&b, t_depth[b] - t_depth[a]);
+		}
+		if(a == b) return a;
```

Comment:
Make it look like C++ code, not C code.
```suggestion
		if(a == b) { return a; }
```